mathesar/utils/explorations.py

```diff
@@ -7,3 +15,115 @@ def get_explorations(database_id):
 
 def delete_exploration(exploration_id):
     Explorations.objects.get(id=exploration_id).delete()
+
+
+def run_exploration(exploration_def, database_id, conn):
+    engine = create_future_engine_with_custom_types(
+        conn.info.user,
+        conn.info.password,
+        conn.info.host,
+        conn.info.dbname,
+        conn.info.port
+    )
+    metadata = get_cached_metadata()
+    base_table_oid = exploration_def["base_table_oid"]
+    initial_columns = exploration_def['initial_columns']
+    params = exploration_def.get('parameters', {})
+    processed_initial_columns = []
+    for column in initial_columns:
+        jp_path = column.get("join_path")
+        if jp_path is not None:
+            join_path = [
+                JoinParameter(
+                    left_oid=i[0][0],
+                    left_attnum=i[0][1],
+                    right_oid=i[1][0],
+                    right_attnum=i[1][1]
+                ) for i in jp_path
+            ]
+        processed_initial_columns.append(
+            InitialColumn(
+                reloid=jp_path[-1][-1][0] if jp_path else base_table_oid,
+                attnum=column["attnum"],
+                alias=column["alias"],
+                jp_path=join_path if jp_path else None
+            )
+        )
+    db_query = DBQuery(
+        base_table_oid=base_table_oid,
+        initial_columns=processed_initial_columns,
+        engine=engine,
+        transformations=exploration_def.get("transformations", []),
+        name=None,
+        metadata=metadata
+    )
+    records = db_query.get_records(
+        limit=params.get('limit', 100),
+        offset=params.get('offset', 0),
+        filter=params.get('filter', None),
+        order_by=params.get('order_by', []),
+        group_by=GroupBy(**params.get('grouping')) if params.get('grouping', None) else None,
+        search=params.get('search', []),
```

Comment:
We may have to do some work to bring the form of these together with the `records.list` form. We should add that to the technical beta catch-up.

Reply:
According to my conversation on the frontend channel, we don't seem to using grouping here, rather we use it from the `transformation` key so I ended up removing it from here.